Img helper size config

allouis opened the pull request:
refs #10181 
Given a theme with a `package.json` with an `image_sizes` object like this:
_(N.B. the `small` key must have at least a height or a width, or both)_
```json
{
  "config": {
    "image_sizes": {
      "small": {
        "width": 300,
        "height": 200
      },
      "wide": {
        "width": 2000
      },
      "tall": {
        "height": 2000
      }
    }
  }
}
```

Using the `img_url` helper in the theme like
```hbs
<img src="{{img_url profile_image}}">
<img src="{{img_url profile_image size="small"}}">
<img src="{{img_url profile_image size="wide"}}">
<img src="{{img_url profile_image size="tall"}}">
<img src="{{img_url profile_image size="unknown-size!!!"}}">
```

Will render to
```html
<img src="/content/images/path/to/profile_image.ext">
<img src="/content/images/size/w300h200/path/to/profile_image.ext">
<img src="/content/images/size/w2000/path/to/profile_image.ext">
<img src="/content/images/size/h2000/path/to/profile_image.ext">
<img src="/content/images/path/to/profile_image.ext">
```

If the `package.json` of the theme does not define the `image_sizes` object or the requested size is not included in the `image_sizes` object, the original image will be served.

allouis committed:
Added "image_sizes" to theme config

allouis committed:
Added activeTheme to helpers proxy

allouis committed:
Updated img-url helper to use const

allouis committed:
Added getImageWithSize helper fn

allouis committed:
Used getImageWithSize fn for img_url helper

allouis committed:
Renamed imageName -> imagePath

allouis committed:
Updated getImageWithSize to use path for sizes

allouis committed:
Updated getImageWithSize middleware to use top-level prefix

allouis committed:
Updated img helper to ensure at least width or height is set for size

allouis committed:
Returned image path when it is absolute

ErisDS commented on core/server/helpers/proxy.js:
```diff
@@ -40,6 +40,8 @@ module.exports = {
         isPrivacyDisabled: config.isPrivacyDisabled.bind(config)
     },
 
+    activeTheme: require('../services/themes/active'),
```
IMO this results in unnecessary tight coupling between the theme service API and the image helper. Keeping in mind that helpers are being moved to Ghost-SDK where possible, not adding new couplings is desirable.

I think we should pass the `image_sizes` config to the helper using the same method we already use for `posts_per-page`, which can be found here: https://github.com/TryGhost/Ghost/blob/f2b44322a3883bdfef500e17ada7c2328d9915d4/core/server/services/themes/middleware.js#L73

ErisDS replied:
Note: @JohnONolan found a bug (of course) where uploading a theme without image sizes still loads image sizes.

Without having double checked the reproduction case or specifics, my first guess is that it happens when uploading a theme without sizes after one that does, and that the config is incorrectly cached between themes?

Hopefully switching out this code will kill 2 birds with 1 stone

allouis replied:
:grimacing: nice bug!

ErisDS commented:
Additional note this PR has the ability to specify only widths/heights which conflicts with the spec and needs confirmation

ErisDS commented:
We've agreed it is preferable to allow either width, height or both.

allouis committed:
Revert "Added activeTheme to helpers proxy"

This reverts commit 3a2accbf2cbb00e2cdd3e9465788c6e6fdd0cb5e.

allouis committed:
Added image_sizes config to themeData

allouis committed:
Updated img_url helper to use image_sizes config

allouis committed:
Renamed sizes -> size

allouis committed:
Destructured width and height config

allouis committed:
Updated path naming to w200h100, w100, h200

allouis commented:
I've updated the PR description to reflect the new url structure i.e `wXhX`, `wX` and `hX`

allouis committed:
Fixed tests for theme middleware

allouis commented on core/server/helpers/img_url.js:
```diff
@@ -17,19 +19,57 @@ module.exports = function imgUrl(attr, options) {
+        return imagePath;
+    }
+
+    if (/https?:\/\//.test(imagePath)) {
```
@ErisDS This check here is to stop us from trying to change images when they're external e.g. unsplash, or from the fixtures. However, when we eventually default to v2 Content API being _the_ api behind rendering themes, this will break - because the Content API is going to serve absolute URL's all the time (I'm pretty sure that's the case?).

ErisDS replied:
Correct, we need to start thinking about images as being a read-only assets API that lives at `[config.url]/content/images`.

At some point, we'll need to take more steps towards implementing it that way, pulling some of the logic out of the current setup.

For now, having utilities that understand this, and can tell us whether an image is coming from the "image API" or something else, should be enough?

allouis replied:
> At some point, we'll need to take more steps towards implementing it that way, pulling some of the logic out of the current setup.

Agreed :)

I'm not 100% sure about the resolution for this however - do we merge it as is? Or wait until it supports the v2 content API?

ErisDS replied:
Unless I'm missing something, this check was added in this PR, IMO the code added here should be compatible with both APIs.

allouis replied:
Ah, right - that's not the case. I was trying to explain that this check _won't_ work with the v2 Content API - because this check essentially says, if the image url is absolute - skip over and don't do anything.

ErisDS replied:
I'm understand that it isn't compatible with the v2 API, but I don't understand this: 

> I'm not 100% sure about the resolution for this however - do we merge it as is? Or wait until it supports the v2 content API?

What is "it" in this suggestion? I assumed the helper, but don't understand who or what we are waiting for - you wrote the code 😬

allouis replied:
:smile: We're waiting for me! Either we wait until next monday and i add something to support Content API v2, or we merge, and raise an issue, to be done either next monday - or some point in the future.

allouis replied:
> IMO the code added here should be compatible with both APIs.

I read this as "currently it should work" rather than "it should be updated so it works" - my :b:

ErisDS replied:
> Either we wait until next monday and i add something to support Content API v2, or we merge, and raise an issue, to be done either next monday 

Either one of those is fine, as long as it is next Monday - given that the v2 API is a high priority item we want to avoid putting extra TODOs in the way of shipping.

allouis committed:
Added support for absolute img url rewriting

no-issue

This is to support the v2 content api, which uses absolute urls for
images. The approach here is to split the url apart, and join it back
together, injecting the size directory in the middle.

allouis committed:
Don't use path join for urls :(

allouis committed:
Add a couple tests

allouis commented:
@ErisDS This has been updated to work with absolute urls when they belong to the blog domain. Which means v2 content api compat :relaxed:
